mountpoint-s3/src/fs.rs

```diff
@@ -603,16 +613,29 @@ where
         ino: InodeNo,
         atime: Option<OffsetDateTime>,
         mtime: Option<OffsetDateTime>,
+        size: Option<u64>,
         _flags: Option<u32>,
     ) -> Result<Attr, Error> {
         tracing::info!(
-            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?}",
+            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?} size {:?}",
             ino,
             _flags,
             atime,
-            mtime
+            mtime,
+            size
         );
-        let lookup = self.superblock.setattr(&self.client, ino, atime, mtime).await?;
+        let setattr_result = self.superblock.setattr(&self.client, ino, atime, mtime).await;
+        let lookup = match (setattr_result, size) {
+            (Ok(lookup), _) => lookup,
+            (Err(InodeError::SetAttrNotPermittedOnRemoteInode(_)), Some(0)) if !self.config.allow_overwrite => {
+                // We want to provide better feedback to users to prompt them to opt-in to file overwrites if it looks like what the application needs.
+                // Instead of complex logic to match `setattr` truncation only, we just check for the error and if the size was set in the request.
+                // If so, we assume its probably a truncation.
+                return Err(
+                    err!(libc::EPERM, "file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it"));
```

Comment:
nit: the immediate error here is still that we do not support `setattr`, we are just guessing that the root cause is an attempt to overwrite. Maybe the message could reflect that, for example something like:

"{original_settattr_error}. If you want to overwrite an existing file, you need to remount ..."

Reply:
With this change the error we get is:
```
WARN setattr{req=11 ino=2 name="twin.txt"}: mountpoint_s3::fuse: setattr failed: file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it
```
It mentions the error is in setattr. And we are checking all the conditions for the root cause to be file overwrite. I think if we want to show both file attributes and file overwrite as root cause, we should not check the condition and keep it for all the `SetAttrOnRemoteInodeNotPermitted` as I was doing it earlier in this PR.